### PR TITLE
[4.0] Fix unit tests and runtime exception issue

### PR DESCRIPTION
**Pull request by yvesh**
### Summary of Changes

Fixes the unit tests in the 4.0-dev branch and an issue in the WebApplication class

### Testing Instructions

Check if travis is green, test that everything is still working and do a code review

### Expected result

RuntimeException is thrown

### Actual result

Fatal error because runtime exception is not thrown

### Documentation Changes Required

None

### CODE REVIEW

File: libraries/src/CMS/Application/WebApplication.php [hunk 1]
```diff
@@ -152,7 +152,7 @@ public static function getInstance($name = null)
 		// Only create the object if it doesn't exist.
 		if (empty(static::$instance))
 		{
-			if (!class_exists($name) && (is_subclass_of($name, '\\Joomla\\CMS\\Application\\WebApplication')))
+			if (!class_exists($name) || !is_subclass_of($name, '\\Joomla\\CMS\\Application\\WebApplication'))
```
Comment (laoneo) [inline libraries/src/CMS/Application/WebApplication.php]: Just for the record, it is like that in staging https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/application/web.php#L206 and 3.8 https://github.com/joomla/joomla-cms/blob/3.8-dev/libraries/src/Joomla/CMS/Application/WebApplication.php#L207. So it should also work in 4.0.

Comment (yvesh) [inline libraries/src/CMS/Application/WebApplication.php]: The test in these two is a positive one - please compare these two again :-) There's a big difference

Comment (wilsonge) [inline libraries/src/CMS/Application/WebApplication.php]: Does this not need to be `!class_exists($name) || class_exists($name) && !is_subclass_of($name, '\\Joomla\\CMS\\Application\\WebApplication')`

File: tests/unit/suites/libraries/joomla/application/JApplicationWebTest.php [hunk 2]
```diff
@@ -327,7 +327,7 @@ public function testFetchConfigurationData($file, $class, $expectsClass, $expect
 	 */
 	public function testGetInstance()
 	{
-		$app = JApplicationWeb::getInstance('JApplicationWebInspector');
+		$app = JApplicationWeb::getInstance('\\JApplicationWebInspector');
```
Comment (laoneo) [inline tests/unit/suites/libraries/joomla/application/JApplicationWebTest.php]: Is this change really needed as it works on 3.8?

**Comment by mbabker** [comment]
Why are we testing that it is a subclass?  I don't remember why I removed it in https://github.com/joomla/joomla-cms/pull/12531 but it really seems fully unnecessary.  There is also no other factory in our API which is doing this type of subclass or interface enforcement.

**Comment by yvesh** [comment]
@mbabker that were the results of Georges merge conflicts - in 3.8 / 3.9 we do it too with a fallback on WebApplication (which does not feel right for me, but this is another topic):

https://github.com/joomla/joomla-cms/blob/3.9-dev/libraries/src/Joomla/CMS/Application/WebApplication.php#L207

**Comment by wilsonge** [comment]
Unless you have an interface that you can rely on I don't understand why we wouldn't do subclass checks. Otherwise we have no clue what methods or properties actually do exist.

**Comment by mbabker** [comment]
It's an arbitrary restriction.  You assume if a factory method can build the object that's being returned, it meets at least whatever is in the doc block.  We don't do these types of checks in any other factory we have.  Why here?

File: tests/unit/suites/libraries/joomla/application/JApplicationWebTest.php (hunk 2 above)
Comment (wilsonge) [inline tests/unit/suites/libraries/joomla/application/JApplicationWebTest.php]: No
